### DISCUSSION TITLE
Larastan (or PHPStan) does not recognised `stdClass` from a query statement

### DISCUSSION THREAD

**Discussion by troccoli** [larastan/larastan#2146] [Q&A] [answered]
Hello everyone,

Before I go into the details, I want to say that I did find a solution, i.e. the code passes level 8, but I was hoping there was a better solution.

I have a seeder to populate a table from a legacy database. For this reason, I don't have, and don't want to, a Model for the legacy data. Instead I'm using the DB facade:

```php
DB::table('sgrdpl_field_data_field_report_programme')
    ->select('field_report_programme_value')
    ->each(fn ($programme) => Programme::query()
        ->insert([
            'name' => $programme->field_report_programme_value,
        ]));
```

On PHPStan level 8 this returns
```
Access to an undefined property object::$field_report_programme_value.
```

The `$programme` is a `stdClass` which should allow dynamic properties. But then Larastan/PHPStan reports that the property is undefined on an `object`. Which is what I really don't understand.

In any case, at first I though I just needed to typehint `$programme` in the closure
```php
DB::table('sgrdpl_field_data_field_report_programme')
    ->select('field_report_programme_value')
    ->each(fn (stdClass $programme) => Programme::query()
        ->insert([
            'name' => $programme->field_report_programme_value,
        ]));
```
And although that fixed the "undefined property", Larastan returned another one
```
Parameter #1 $callback of method Illuminate\Database\Query\Builder::each() expects callable(object, int): mixed, Closure(stdClass): int given
```
Normally, I would create a specific Builder for my model, but I don't have a model here. So I tried with another typehint:
```php
DB::table('sgrdpl_field_data_field_report_programme')
    ->select('field_report_programme_value')
    ->each(fn (object $programme) => Programme::query()
        ->insert([
            'name' => $programme->field_report_programme_value,
        ]));
```
but then resulted in the same "undefined property" error. I even tried to add `object` to the `universalObjectCratesClasses` array in my configuration.

I read about object shape and after a bit of trial and error the following works
```php
DB::table('sgrdpl_field_data_field_report_programme')
    ->select('field_report_programme_value')
    ->each(function ($p) {
        /** @var object{field_report_programme_value: string} $programme */
        $programme = (object) $p;
        Programme::query()->insert([
            'name' => $programme->field_report_programme_value,
        ]);
    });
```

As I said, that works. But it is a bit ugly. Besides, this is a simple example, where I select only one field. But I have other seeders which instead select a dozen fields and declaring an object shape it's a lot of lines (I know I can split the declaration into multiple lines).

I was wondering if there is a better solution.

**Comment by szepeviktor** [comment]
It can be simplified
```php
DB::table('sgrdpl_field_data_field_report_programme')
    ->select('field_report_programme_value')
    ->each(function ($programme) {
        /** @var object{field_report_programme_value: string} $programme */

        ->insert([
            'name' => $programme->field_report_programme_value,
            ]);
    });
```

**Reply by troccoli** [reply]
Mmm, that only gets rid of the casting, and it doesn't work if I use the arrow function (which I did try).

My question was about getting rid of the whole object shape if possible.

**Comment by rudiedirkx** [comment]
Why use `->each(Closure)` at all? Why not just use `foreach`? `QueryBuilder->get()` returns `array<stdClass>` since recently.

**Reply by troccoli** [reply]
Because to use a `foreach` I first need to get all records. Instead the `each` method on a query retrieves the record in chunks and then processes them one by one.

**Reply by rudiedirkx** [reply]
Wait really? I learned something! I always use `lazy()` (or sometimes `chunk()`) for that, and then a `foreach`. Generators ftw. I don't think `lazy()` will return a `Generator<stdClass>` though, so that doesn't help you. I'll go back into my hole now.

**Reply by troccoli** [reply]
I was trying to find the use of the `each()` method in that way in the Laravel documentation, and I cannot. I only found this use
```php
Flight::where('departed', true)
    ->lazyById(200, column: 'id')
    ->each->update(['departed' => false]);
```
which, I think, uses the the `each` as high-order method, so basically that's equivalent to
```php
Flight::where('departed', true)
    ->lazyById(200, column: 'id')
    ->each(fn ($flight) => $flight->update(['departed' => false]));
```
I don't think the `lazyById` here makes any difference, btw

**Comment by canvural** [comment, marked as answer]
The `object` is coming from [here](https://github.com/laravel/framework/blob/abf823f36b5cb31b910cbbb14225dccfeeb52263/src/Illuminate/Database/Query/Builder.php#L37) And indeed it's wrong, it should be `stdClass` And the undefined property error comes from PHPStan itself and it's correct. `object` can be any object and PHPStan doesn't know the properties on it.

Feel free to send PR to Laravel, and you can also update the stubs here if you want.

**Comment by Roardom** [comment]
This was recently changed: https://github.com/laravel/framework/pull/58150
